[InstCombine] Compare `icmp inttoptr, inttoptr` values directly

citymarina opened the pull request:
InstCombine already has some rules for `icmp ptrtoint, ptrtoint` to drop the casts and compare the source values. This change adds the same for the reverse case with `inttoptr`.

fhahn commented on llvm/test/Transforms/InstCombine/icmp.ll:
```diff
@@ -5326,3 +5326,22 @@ define i1 @pr94897(i32 range(i32 -2147483648, 0) %x) {
+; CHECK-LABEL: @test_icmp_inttoptr(
+; CHECK-NEXT:    [[CMP:%.*]] = icmp eq i64 [[X:%.*]], [[Y:%.*]]
+; CHECK-NEXT:    ret i1 [[CMP]]
+  %xptr = inttoptr i64 %x to ptr
```
This probably also needs negative tests (including source integer types that != pointer width, only one op is `inttoptr`) and possibly also some cases where `inttoptr` have multiple users?

citymarina replied:
Good point about the multiple users. Regarding the varying integer widths, I had tried writing such tests but the values ended up getting a `trunc` or `zext` to pointer width anyway, so they still passed through this code.

nikic replied:
inttoptr with mismatching sizes gets canonicalized, but it's still good to add a test for it.

nikic commented:
Please see https://llvm.org/docs/InstCombineContributorGuide.html, esp. the bit about pre-committing tests.

nikic commented on llvm/test/Transforms/InstCombine/icmp.ll:
```diff
@@ -5326,3 +5326,22 @@ define i1 @pr94897(i32 range(i32 -2147483648, 0) %x) {
+define i1 @test_icmp_inttoptr(i64 %x, i64 %y) {
+; CHECK-LABEL: @test_icmp_inttoptr(
+; CHECK-NEXT:    [[CMP:%.*]] = icmp eq i64 [[X:%.*]], [[Y:%.*]]
+; CHECK-NEXT:    ret i1 [[CMP]]
```
Did you manually write these? The generated output should have one more `;`.

citymarina commented:
Oh wow, that contributor guide looks really thorough. Thank you for writing it!

dtcxzyw reviewed (commented):
My last attempt https://github.com/llvm/llvm-project/pull/77832 broke hwasan buildbot. @vitalybuka Can you check this patch?

citymarina committed:
[InstCombine] Add tests for fold of `icmp inttoptr, inttoptr`. NFC

citymarina commented on llvm/test/Transforms/InstCombine/icmp-inttoptr.ll:
```diff
@@ -71,8 +63,7 @@ define i1 @inttoptr_size_mismatch(i200 %x, i9 %y) {
 ; CHECK-NEXT:    [[TMP1:%.*]] = trunc <2 x i200> [[X:%.*]] to <2 x i64>
-; CHECK-NEXT:    [[XPTR:%.*]] = inttoptr <2 x i64> [[TMP1]] to <2 x ptr>
-; CHECK-NEXT:    [[CMP:%.*]] = icmp eq <2 x ptr> [[XPTR]], <ptr inttoptr (i9 42 to ptr), ptr inttoptr (i9 123 to ptr)>
+; CHECK-NEXT:    [[CMP:%.*]] = icmp eq <2 x i64> [[TMP1]], <i64 42, i64 123>
```
I had to change to `max-iterations=2` because this test case goes through an intermediate stage:
```
<ptr inttoptr (i9 42 to ptr), ptr inttoptr (i9 123 to ptr)>
-->
<i64 ptrtoint (ptr inttoptr (i9 42 to ptr) to i64), i64 ptrtoint (ptr inttoptr (i9 123 to ptr) to i64)>
-->
<i64 42, i64 123>
```
Is that ok?

nikic replied:
Replacing ConstantExpr::getPtrToInt with use of IRBuilder should fix this.

citymarina replied:
I'm new in this area and could use some more explicit guidance. I _think_ you are suggesting to use something like `return replaceInstUsesWith(ICmp, Builder.CreateICmp(ICmp.getPredicate(), Op0Src, ...))` but I am not clear about what goes in the `...`. Do I need to crack open the `ConstantAggregate` and do something with that?

nikic replied:
No, I mean to replace `NewOp1 = ConstantExpr::getPtrToInt(RHSC, SrcTy);` with `NewOp1 = Builder.CreateIntToPtr(RHSC, SrcTy);`

The difference is that the IRBuilder will perform DataLayout-aware constant folding.

You could also directly invoke the ConstantFoldCastOperand() API, but going through IRBuilder is preferred in this context.

citymarina replied:
> But even with this, the test case still goes through the intermediate state of `<i64 ptrtoint (ptr inttoptr ...` so it doesn't reach fixpoint in one iteration.

@nikic ping; how do you recommend to proceed? Thanks.

nikic replied:
I looked into it, and the problem is that ConstantFoldCastInst expects the operand of the ptrtoint to be an inttoptr expression, but here it is a vector of inttoptr. The fold only works after the ptrtoint has been applied element-wise to the vector elements. We should probably fix the constant folding API to make this work directly, but for now you can work around it like this: `ConstantFoldConstant(ConstantExpr::getPtrToInt(RHSC, SrcTy), DL)`

citymarina replied:
Thanks for the pointer!

citymarina commented:
> Oh wow, that contributor guide looks really thorough. Thank you for writing it!

I've read through the guide carefully and hopefully this revision better meets the expectations. I appreciate the clear instructions!

nikic commented on llvm/test/Transforms/InstCombine/icmp-inttoptr.ll:
```diff
@@ -0,0 +1,101 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -passes='instcombine<max-iterations=2>' -S | FileCheck %s
+
+target datalayout = "e-p:64:64:64-p1:16:16:16-p2:32:32:32-p3:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v64:64:64-v128:128:128-a0:0:64-s0:64:64-f80:128:128-n8:16:32:64"
```
The data layout should not be needed?

citymarina commented:
Ping

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -6081,12 +6081,12 @@ Instruction *InstCombinerImpl::foldICmpWithCastOp(ICmpInst &ICmp) {
+      IntTy = cast<VectorType>(IntTy)->getElementType();
     }
-    return DL.getPointerTypeSizeInBits(SrcTy) == DestTy->getIntegerBitWidth();
+    return DL.getPointerTypeSizeInBits(PtrTy) == IntTy->getIntegerBitWidth();
```
Why the variable rename?

citymarina replied:
The existing code for `ptrtoint` calls `CompatibleSizes(SrcTy, DestTy)`. The code that I added for `inttoptr` calls with the arguments in the reverse order: `CompatibleSizes(DestTy, SrcTy)`. It felt awkward to pass a `Dest` to a function expecting `Src` and vice versa. I figured that anyway the more important distinction is which is the int and which is the pointer.

goldsteinn commented on llvm/test/Transforms/InstCombine/icmp-inttoptr.ll:
```diff
@@ -0,0 +1,99 @@
+  %xptr = inttoptr <2 x i200> %x to <2 x ptr>
+  %cmp = icmp eq <2 x ptr> %xptr, inttoptr (<2x i9> <i9 42, i9 123> to <2 x ptr>)
+  ret <2 x i1> %cmp
+}
```
For the two size mismatch cases I think the tests would be improved if you only had one mismatch.

I.e i200 -> ptr on LHS in the scalar test and i9 - ptr on RHS in the vec test.

citymarina replied:
Fixed, thanks.

nikic commented:
I think the open question here is:

> My last attempt #77832 broke hwasan buildbot. @vitalybuka Can you check this patch?

But maybe we can just land this and see what happens.

citymarina commented:
> I think the open question here is:
> 
> > My last attempt #77832 broke hwasan buildbot. @vitalybuka Can you check this patch?
> 
> But maybe we can just land this and see what happens.

I interpreted @vitalybuka's removing themself from the review to indicate no plans to test this, but if that was incorrect let me know.

citymarina committed:
[InstCombine] Compare `icmp inttoptr, inttoptr` values directly

InstCombine already has some rules for `icmp ptrtoint, ptrtoint` to
drop the casts and compare the source values. This change adds the
same for the reverse case with `inttoptr`.

XChy reviewed (approved):
Look reasonable to me

dtcxzyw commented:
> I think the open question here is:
> 
> > My last attempt #77832 broke hwasan buildbot. @vitalybuka Can you check this patch?
> 
> But maybe we can just land this and see what happens.

Agree.

dtcxzyw reviewed (approved):
LG

nikic commented:
Looks like the hwasan build bot with this commit passed: https://lab.llvm.org/buildbot/#/builders/55/builds/2229

vitalybuka commented:
> My last attempt #77832 broke hwasan buildbot. @vitalybuka Can you check this patch?

Sorry, I don't expert in this code, so I removed myself, assuming I am accidentally here. I didn't noticed testing request.
In general if you  confident, it's fine to land and wait, because even for owner of the bot testing there is some hassle.
